### DISCUSSION TITLE
[http] require `accept` header?

### DISCUSSION THREAD

**Discussion by glasser** [graphql/defer-stream-wg#48] [General]
I think this WG is mostly for the underlying spec changes, not the HTTP transport, so rather than have discussion about this here, I'm just calling out an issue on the `graphql-over-http` repository. I've recently added a comment to this old issue with my thoughts. The tl;dr: I think clients should send a header like `accept: multipart/mixed, application/json` and servers should not send `multipart/mixed` responses to clients that don't ask for it, and that they should send errors to those clients if they use an incremental delivery directive.

Issue comment: https://github.com/graphql/graphql-over-http/issues/167#issuecomment-1222741926

**Comment by maraisr** [comment]
agree the content type should convey its responding with multipart contents! In fact meros will not process the response as multipart if the server didn't send it ~> https://github.com/maraisr/meros/blob/e63608b1926e7f8c7ab217b72bd9b425b274c305/src/browser.ts#L97

but vaguely remember there being chatter about the `accept` header, as in the presence of `@defer` should exist in the same space as the person making the call, and generally the network doesn't directly talk to the executor. So you'd have to strip the directives before giving it to the executor. 

So 💯  about response types, but 👀  about accept header — imo. Think of it like, the user decides if it wants to defer, not the server. So if you defer, by virtue you accept as well.

my 2c

**Reply by glasser** [reply]
I'm not sure I follow. You start by saying you agree and that your implementation matches my suggestion... but the rest seems to disagree?

My concern is that if you run, say, a version of GraphiQL that doesn't actually know how to parse multipart/mixed, then the naive user typing a `@defer` operation into the UI should ideally get a nice error about the transport not supporting `@defer` instead of a confusing JSON parse error. So no stripping required.

**Reply by maraisr** [reply]
ill edit to reflect that I agree about the content type, not about the `accept` header. The only time a schema would validate the `@defer` directive is supported, would be if the server accepted that as a valid directive. 

If the client doesn't know how to deal with a schema that supports it, that is the same as firing `subscription` queries to a schema that supports it, but no ws on the client.

**Reply by glasser** [reply]
I think servers are generally capable of giving a good error if `subscription` operations are sent on a transport that doesn't support them (eg, a standard HTTP POST without the websocket upgrade). I'm trying to do the same thing by making it easy for servers to give a good errors if `@defer` is sent without giving a hint that you will be able to understand the response.

**Reply by maraisr** [reply]
The thing is executors and transports are specifically kept seperate, so to tie them together is a bad move.

At best, and not specifically a spec move imo — but a user land "recommendation" would be for the network to set an executor flag, say `incrementalDelivery: false` if the network deems necessary. Which afaik can be done today. As to say "if accept header not multipart, then executor you better not yield".

**Reply by glasser** [reply]
I'm not sure I follow your argument. In any case, can we take this over to the issue on the graphql-over-http repository as I suggested? This seems in line with your point that transport discussion should be kept separate from execution discussion.

**Reply by maraisr** [reply]
You don't seem to follow many things, maybe my bad. Quite frankly, accept headers and multipart guarantees is not something the executer should care about. If a server wants to disable that feature based on some signal (the lack of accept header), then so be it. But not something to be specified in the spec.

**Reply by glasser** [reply]
The spec I am talking about is [this one](https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md) which is all about headers and multipart formatting. (And it's an RFC for updating a [larger spec](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md) which has plenty of thoughts about the Accept header specifically.)

**Reply by maraisr** [reply]
sounds to me like you're wanting to chat about transport specifics, as both of those links link away from this repo.

@robrichard do we have chatter about the transport here?
